File: TrapRainWater/TrapRainWater.cpp
// TrapRainWater.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <vector>

size_t GetNextHighest(const std::vector<int>& heights, size_t begin)
{
    int height = heights[begin++];
    while (begin != heights.size())
    {
        if (heights[begin] < heights[begin - 1])
            return begin - 1;
    }

    return begin - 1;
}

bool NoHigherLeft(const std::vector<int>& heights, size_t begin)
{
    for (auto i = begin + 1; i < heights.size(); ++i)
        if (heights[i] > heights[begin])
            return false;
    return true;
}

size_t GetNextTrap(const std::vector<int>& heights, size_t begin)
{
    int highest = heights[begin];
    int lowest = heights[begin];

    while (++begin != heights.size())
    {
        if (heights[begin] >= highest)
            return begin;

        if (heights[begin] > heights[begin - 1])
        {
            if (NoHigherLeft(heights, begin))
                return begin;
        }

        if (heights[begin] < lowest)
            lowest = heights[begin];
    }

    return -1;
}

int VolumeOfTrappedRainWater(const std::vector<int>& heights)
{
    size_t endHeight = 0;
    size_t startHeight = 0;
    int volume = 0;

    while (startHeight < heights.size())
    {
        startHeight = GetNextHighest(heights, endHeight);
        if (startHeight == heights.size()-1)
            break;

        endHeight = GetNextTrap(heights, startHeight);
        if (endHeight == -1)
        {
            endHeight = startHeight + 1;
            continue;
        }

        int trappingHeight = std::min(heights[startHeight], heights[endHeight]);
        for (++startHeight; startHeight < endHeight; ++startHeight)
        {
            if (heights[startHeight] < trappingHeight)
                volume += trappingHeight - heights[startHeight];
        }
    }

    return volume;
}

int main()
{
    while (true)
    {
        int count = 0;
        std::cout << "Number of elements in the array: ";
        std::cin >> count;
        if (count == 0)
            break;

        std::vector<int> vec(count);
        std::cout << "The numbers of the sorted array: ";
        for (int i = 0; i < count; ++i)
            std::cin >> vec[i];

        std::cout << "The trapped rain volume is: " << VolumeOfTrappedRainWater(vec) << std::endl << std::endl;
    }
}

File: Go/TrapRainWater.go
package main

import "fmt"

func FindNextHighest(heights []int, index int) int {
	for index = index + 1; index < len(heights); index++ {
		if heights[index] < heights[index-1] {
			return index - 1
		}
	}

	return -1
}

func NoHigherLeft(heights []int, index int) bool {
	for i := index + 1; i < len(heights); i++ {
		if heights[i] > heights[index] {
			return false
		}
	}

	return true
}

func GetTrap(heights []int, index int) int {
	lowest := heights[index]
	highest := heights[index]

	for index = index + 1; index < len(heights); index++ {
		if heights[index] >= highest {
			return index
		}
		if heights[index] > heights[index-1] {
			if NoHigherLeft(heights, index) {
				return index
			}
		}

		if heights[index] < lowest {
			lowest = heights[index]
		}
	}

	return -1
}

func FindNextTrap(heights []int, start int) (int, int) {
	start = FindNextHighest(heights, start)
	if start == len(heights)-1 {
		return start, -1
	}
	highest := GetTrap(heights, start)

	return start, highest
}

func VolumeOfTrappedRainWater(heights []int) int {
	start := 0
	volume := 0
	end := 0
	for start < len(heights) {
		if end == len(heights)-1 {
			break
		}

		start, end = FindNextTrap(heights, end)
		if end == -1 {
			end = start + 1
			continue
		}

		trappingHeight := min(heights[start], heights[end])
		for start++; start < end; start++ {
			if heights[start] <= trappingHeight {
				volume += (trappingHeight - heights[start])
			}
		}
	}

	return volume
}

func main() {
	for {
		fmt.Print("Number of heights: ")
		count := 0
		fmt.Scan(&count)
		if count == 0 {
			break
		}

		fmt.Print("Please enter the heights: ")
		heights := make([]int, count)
		for i := 0; i < count; i++ {
			fmt.Scan(&heights[i])
		}

		fmt.Print("The volume of trapped water is ")
		fmt.Println(VolumeOfTrappedRainWater(heights))
	}
}
